src/Units/UnitContext.java

```diff
@@ -97,9 +118,84 @@ public void setEnvironment(Environment input)
       terrainStars = env.terrainType.getDefLevel();
   }
 
+  public void setWeapon(WeaponModel w)
+  {
+    weapon = w;
+    if( null != w )
+    {
+      rangeMin = w.rangeMin;
+      rangeMax = w.rangeMax;
+      // Build a scratch UnitContext because we don't trust UnitMods with our state
+      UnitContext uc = new UnitContext(this);
+      for( UnitModifier mod : mods )
+        mod.modifyAttackRange(uc);
+      rangeMin = uc.rangeMin;
+      rangeMax = uc.rangeMax;
```

Comment:
Mmmm, Pungent.

Reply:
The alternatives (while also being reasonably defensive) are:
Split `UnitContext` into a bunch of separate classes for different contexts (messy, probably lots of duplication, and could end up looking really dumb depending on CO design)
Access protection (requires a bunch of interfaces and functions I don't want to write)
Make state immutable, so we'd be making a copy anyway

Reply:
Also what do you mean that we don't trust UnitMods with the state? Isn't their whole point that they modify state?

Reply:
Their whole point is modifying state, yes... but they _can_ modify more fields than just the ones they're supposed to.
Perhaps what I should be doing is enforcing "If a modifier changes something unrelated to what it's supposed to be changing, that's a bug".

Reply:
If we can't trust the unit modifiers to only change what they should, we have bigger problems. There should ideally be a unit test per modifier to ensure it always and only does what it should, and then make sure we aren't double-applying. This kind of defensive construct feels... bad.

> If a modifier changes something unrelated to what it's supposed to be changing, that's a bug

Yes by definition?

Reply:
> There should ideally be a unit test per modifier to ensure it always and only does what it should

Most of them fall into the level of "trivial", in my book. Adding to or subtracting from a single number is about all they do.
> Yes by definition?

Okay. Sounds good.